토큰 엔티티 필드 조회 기능 추가

kayoung0 opened the pull request:
## Description
1. TokenController 추가
2. TokenService 추가
3. TokenRepository 추가
4. EventDto 추가
5. TokenDto 추가
6. application-local.yml 추가

## Changes
**1. TokenController**
token: TokenDto를 프론트와의 통신양식에 따라 반환
- @GetMapping(value="/token/{id}")일 때 tokenService의 getToken 함수를 호출합니다.
- getToken 호출 결과물인 TokenDto를 ApiResponse의 data에 실어 반환합니다.

**2. TokenService**
getToken: token id를 가지는 token entity를 찾은 후 TokenDto를 반환합니다.
- 매개변수는 token의 id입니다. 
- repository의 findById를 호출하여 해당 id를 가지는 token entity를 받아옵니다.
- id를 가지는 token entity가 존재하지 않는다면 knuException을 발생시켜  400001 code와 에러 메시지를 반환합니다.

**3. Event, Token entity의 desc필드를 description로 변경**
- desc가 예약어라서 event, token table이 생성되지 않습니다. desc를 description으로 변경했습니다.

**4. TokenDto**
- fromEntity: static method입니다. 매개변수로 Token을 받고 TokenDto로 변환한 후 반환합니다.
- toEntity: TokenDto를 Token으로 변환한 후 반환합니다.

**5. application-local.yml 추가**
- mysql의 url, username, password 설정값이 있습니다.
- username과 password는 환경변수로 관리합니다.


## Test Checklist
- token, event entity에 해당하는 테이블 생성 확인
- token, event table이 event table의 id 컬럼을 기준으로 join됨을 확인
- tokenId에 해당하는 토큰이 있는 경우, /token/{tokenId}로 접속 시 ApiResponse.ok 반환 확인
- tokenId에 해당하는 토큰이 없는 경우, TokenService 클래스에서 knuException 예외 발생 확인

PPakSang committed:
Add: Event, Token Entity 추가
NFT 발행 및 조회에 사용되는 NFT 메타데이터를 저장하는 Entity 를 추가합니다.

Event: NFT 콜렉션 정보
Token: 발행 완료 NFT 의 메타데이터 정보

kayoung0 committed:
feat: 토큰 정보 조회

/token/{tokenId} 접근하면 token 정보를 Json으로 생성하고 문자열로 반환

kayoung0 committed:
Merge dev into feature/token and resolve conflicts

kayoung0 committed:
chore: .gitignore 적용

kayoung0 committed:
fix: Event, Token entity의 desc 필드에 @Column(name="description") 추가

desc가 예약어라서 테이블이 만들어지지 않았기 때문에 column 이름을 "description"으로 변경

kayoung0 committed:
chore: 폴더 구조 변경

Token 관련 controller, dto, repository, service를 token 디렉토리로 이동시킴

chanwooDev committed:
feat: 커스텀 exception에 별도의 메세지로 설정하여 반환하도록 수정 (#6)

chanwooDev commented on src/main/java/com/knu/ntttt_server/token/controller/TokenController.java:
```diff
@@ -0,0 +1,27 @@
+    ObjectMapper objectMapper = new ObjectMapper();
+    return objectMapper.writeValueAsString(tokenDto);
+  }
+
```
https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/2
프론트와 통신하는 form양식을 정해놓았습니다. 위의 사항대로 처리해주세요!

kayoung0 replied:
@kayoung0
[fix: Response, RuntimeException 반환 양식 준수](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/d98d86dffcee5424b585a347ee1cd900505f789d) 
[d98d86d](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/d98d86dffcee5424b585a347ee1cd900505f789d)
400_001 "cannot find token" ResultCode 추가

kayoung0 replied:
**response**
{"result":{"code":200001,"message":"success"},"data":{"id":2,"event":{"id":1,"name":"sm town concert","publisher":"SM","quantity":3,"description":"2022년 sm town concert입니다."},"seq":2,"imgUrl":"/artist/winter.jpg","price":35600,"description":"윈터가 도깨비불의 포인트 안무를 선보이고 있다.","nftId":43254321,"paymentState":"SOLD_OUT","owner":"kwangya"}}
**knuException**
{"result":{"code":400001,"message":"cannot find token"},"data":null}

chanwooDev commented on src/main/java/com/knu/ntttt_server/token/controller/TokenController.java:
```diff
@@ -0,0 +1,27 @@
+public class TokenController {
+
+  private TokenService tokenService;
+  public TokenController(TokenService tokenService){
```
1. 변수와 함수 사이 엔터 부탁드려요 ,(함수랑 함수사이도 엔터)
2. e){ -> ) {  스페이스 한번

kayoung0 replied:
@kayoung0
[comment: TokenController, TokenService 주석 추가](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/2f22506ee6e8eb4964ecfedfa6aa3332bbbd1d4b)

@kayoung0
[style: 코드 컨벤션에 맞게 수정](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/b5b445995bd1fd0ab8833ba49d69723a6edc2e49)

PPakSang commented on src/main/java/com/knu/ntttt_server/token/dto/TokenDto.java:
```diff
@@ -0,0 +1,51 @@
+    this.owner = owner;
+  }
+
+  public Token toEntity(Event event, Long seq, String imgUrl, Long price, String description, Long nftId, PaymentState paymentState, String owner){
+    Token token = Token.builder()
+        .event(event)
+        .seq(seq)
+        .imgUrl(imgUrl)
+        .price(price)
+        .desc(description)
+        .nftId(nftId)
+        .paymentState(paymentState)
+        .owner(owner)
+        .build();
+
+    return token;
+  }
```
Dto 를 Entity 로 변환하는 메소드를 작성하신 것 같습니다.
이 경우에는 파라미터로 필드값을 받아오는게 아닌, Dto 의 멤버변수(this.someField) 를 통해 Entity 객체를 생성하도록 작성하는게 나아보이네요!

Event 클래스의 경우 Token 의 연관관계 Entity 로 Dto 를 생성하는 시점에 초기화가 안되어 있을 수도 있고, JPA 에 의해 예측 불가능한 쿼리가 생성될 위험이 있습니다. 이에 대한 고민이 필요합니다 (Event 필드 삭제 혹은 Event 에서 필요한 필드값을 외부에서 받아오는 방식 등)

kayoung0 replied:
@kayoung0
[fix: Token, Event entity의 desc 필드를 description으로 변경](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/de091023bd44745aaca64fb1e12e543670f604d7)

[fix: Token, Event entity의 toEntity method 수정](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/97655f5d633d930bd568c7713cc05b8f0b85daba)

TokenDto -> Dto 의도에 맞는 네이밍 고려해야 함

PPakSang commented on src/main/resources/application.yml:
```diff
@@ -1,2 +1,18 @@
+logging:
+  level:
+    org.hibernate.SQL: debug
+    org.hibernate.type: trace
```
외부에 공개되면 안되는 정보(url, username, password 등등)은 환경변수로 관리해주세요!

```yml
url: ${URL}
```

chanwooDev replied:
해당 파일은 application-local.yml로 빼야할 것 같아요
로컬이라 url은 둬도 될 것 같고 password, username은 환경변수로 해도 될 것 같습니다.

kayoung0 replied:
@kayoung0
[refactor: application-local.yml 추가](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/d9890ce7acde75a0daee25225cf0331a39b5480d)
[d9890ce](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/d9890ce7acde75a0daee25225cf0331a39b5480d)
application.yml의 spring, logging 부분을 application-local.yml로 이동
database username, password를 환경변수로 관리

PPakSang commented on src/main/java/com/knu/ntttt_server/token/dto/EventDto.java:
```diff
@@ -0,0 +1,33 @@
+      .build();
+
+    return event;
+  }
```
Token toEntity 부분과 같은 의견입니다.

chanwooDev commented on src/main/java/com/knu/ntttt_server/token/dto/TokenDto.java:
```diff
@@ -0,0 +1,51 @@
+        .nftId(nftId)
+        .paymentState(paymentState)
+        .owner(owner)
+        .build();
```
어떻게 쓰는지 예제를 한번 찾아보고 적용해보세요! (구글 검색)

chanwooDev commented on src/main/resources/application.yml:
```diff
@@ -1,2 +1,18 @@
+logging:
+  level:
+    org.hibernate.SQL: debug
+    org.hibernate.type: trace
```
EOF 가 없네요 추가해주세용

kayoung0 replied:
@kayoung0
[Chore: EOF 추가](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/dadc20c65a93741ce2d0d100e1c5f0447b2e54f0) 
[dadc20c](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/dadc20c65a93741ce2d0d100e1c5f0447b2e54f0)
application.yml, application-local.yml EOF 추가

chanwooDev commented on src/main/java/com/knu/ntttt_server/token/service/TokenService.java:
```diff
@@ -0,0 +1,35 @@
+
+  public TokenDto getToken(Long tokenId){
+    Optional<Token> tokenWrapper = tokenRepository.findById(tokenId);
+    Token token = tokenWrapper.get();
```
데이터가 존재하지 않는 ID로 요청하면 어떻게 될까요?

https://www.latera.kr/blog/2019-07-02-effective-optional/
참고해서 Exception 잡아주세요! (해당 id가 없는 경우)

chanwooDev replied:
[fix: Response, RuntimeException 반환 양식 준수](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/d98d86dffcee5424b585a347ee1cd900505f789d)

chanwooDev commented on src/main/java/com/knu/ntttt_server/token/service/TokenService.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Optional;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TokenService {
```
@RequiredArgsConstructor 을 사용해보세요!

chanwooDev replied:
@kayoung0 

예시예여 
[fix: TokenService 생성자 삭제](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/b0047c342d29867034b9263258b19cb35b42375f)

kayoung0 replied:
@kayoung0
[fix: TokenService 생성자 삭제](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/b0047c342d29867034b9263258b19cb35b42375f)

kayoung0 committed:
refactor: application-local.yml 추가

application.yml의 spring, logging 부분을 application-local.yml로 이동
database username, password를 환경변수로 관리

kayoung0 committed:
Chore: EOF 추가

application.yml, application-local.yml EOF 추가

kayoung0 committed:
fix: Response, RuntimeException 반환 양식 준수

400_001 "cannot find token" ResultCode 추가

PPakSang committed:
서버 스펙 NFT 발행, 전송, 조회 기능 추가 (#4)

* Add: Web3j Dependency 및 설정 정보(환경 변수) 추가

* Add: ContractUtils 추가
ContractUtils: Web3j 를 통해 스마트컨트랙트와 통신을 수행

OwnerAddr, ContractAddr 을 통해 스마트컨트랙트 내 함수를 호출합니다.
sendTransaction: 가스를 소모하는 함수 호출 (삽입, 수정)
call: 가스를 소모하지 않는 함수 호출 (조회)
getReceipt: 비동기 트랜잭션의 결과(로그 데이터 사용) 조회

코드 상세 설명은 주석확인 부탁드립니다.

* Add: NftService Interface 추가

서버 내 NFT 관련 기능 명세를 추가합니다
mintNft: 서버 명의의 nft 발행
transferNft: 서버 명의의 nft 전송
queryNft: nft 조회

* Add: NftServiceImpl, Dto 추가

서버 스펙의 NFT Dto 를 추가합니다. (추후 변경 가능 영역)
imgUrl: NFT 이미지 주소
desc: NFT 상세 정보

서버가 배포한 스마트컨트랙트와 통신해 NFT 관련 기능을 수행하는 NftService 구현체를 추가합니다.

mintNft: NFT 생성 요청 데이터를 바탕으로 NFT 를 발행하고, 발행된 NFT 의 ID 를 반환합니다.
- 비동기로 트랜잭션이 요청되기 때문에 추후에 시간을 두고 배치작업으로 응답 결과를 받아와야하는 영역입니다. 현재는 (데이터 처리량이 많지 않아 곧바로 처리되기 때문에) 바로 트랜잭션 결괏값을 받아오도록 코드를 작성했습니다

- 결괏값(NFT ID) 을 받아오는 로직 같은 경우에는 openzepplin storage(스마트컨트랙트) 에서 생성되는 로그데이터를 참고해 직접 접근하는 방식으로 스마트컨트랙트 코드와 결합도가 다소 높습니다.

transferNft: 서버 명의의(서버에서 발행한) NFT 를 다른 주소로 전송합니다.
- 서버가 보유하고 있지 않은 NFT 를 타인에게 전송할 시 예외(Transfer Failed) 를 발생시킵니다

queryNft: NFT ID 값을 바탕으로 NFT 데이터를 반환합니다.
- 서버 스펙의 NFT 필드를 보유하고 있지 않은 경우 예외(JSON Parsing Error) 를 발생시킵니다

* Chore: EOF 추가

* Chore: from -> of 메소드 시그니처 변경 (다중 파라미터), 2중 스페이스바 제거

kayoung0 committed:
fix: Token, Event entity의 desc 필드를 description으로 변경

kayoung0 committed:
fix: Token, Event entity의 toEntity method 수정

kayoung0 committed:
comment: TokenController, TokenService 주석 추가

kayoung0 committed:
style: 코드 컨벤션에 맞게 수정

kayoung0 committed:
fix: TokenService 생성자 삭제

chanwooDev commented on src/main/java/com/knu/ntttt_server/token/controller/TokenController.java:
```diff
@@ -0,0 +1,31 @@
+    TokenDto tokenDto = tokenService.getToken(tokenId);
+    return ApiResponse.ok(tokenDto);
+  }
+
+
```
의미없는 엔터가 있네요!

kayoung0 replied:
@kayoung0
[comment: TokenController, TokenService 주석 요약, 의미없는 엔터 제거](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/52fce8e53ca24b5e01a2eb9a3ce539584dce3d2f)

chanwooDev commented on src/main/java/com/knu/ntttt_server/token/service/TokenService.java:
```diff
@@ -0,0 +1,38 @@
+  public TokenDto getToken(Long tokenId) {
+    Token token = tokenRepository.findById(tokenId)
+        .orElseThrow(() -> new KnuException(ResultCode.BAD_REQUEST));
+    return TokenDto.builder()
+        .id(token.getId())
+        .description(token.getDescription())
+        .event(token.getEvent())
+        .imgUrl(token.getImgUrl())
+        .nftId(token.getNftId())
+        .paymentState(token.getPaymentState())
+        .price(token.getPrice())
+        .seq(token.getSeq())
+        .owner(token.getOwner())
+        .build();
```
정적 팩토리메서드 패턴을 만들어 TokenDto안에 함수로 만들면 좋을 것같네요

```java
public static TokenDto from(Token token)
return new ~~~
```

kayoung0 replied:
@kayoung0
[Refactor: TokenDto에 static 메서드 fromEntity() 생성](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/08ff560274ea7ea6f271018e68eb3cae01e5895e)

kayoung0 committed:
comment: TokenController, TokenService 주석 요약, 의미없는 엔터 제거

kayoung0 committed:
Refactor: TokenDto에 static 메서드 fromEntity() 생성

chanwooDev commented on src/main/java/com/knu/ntttt_server/token/service/TokenService.java:
```diff
@@ -20,16 +20,6 @@ public class TokenService {
   public TokenDto getToken(Long tokenId) {
     Token token = tokenRepository.findById(tokenId)
         .orElseThrow(() -> new KnuException(ResultCode.BAD_REQUEST));
-    return TokenDto.builder()
```
명확한 메세지를 넣어서 이유를 알게해주는게 좋을 것 같네유

```java

KnuException(ResultCode.BAD_REQUEST, "해당하는 토큰의 id가 존재하지 않습니다."))
```

kayoung0 replied:
[fix: 토큰이 존재하지 않을 때 발생하는 예외에 메시지 추가](https://github.com/NTTTT-NFT-Ticket-Trading-Town/NTTTT-BE/pull/5/commits/7642e5dd4779517aff3ef359229e1bcf63bb0b2d)

{"result":{"code":400001,"message":"해당하는 토큰의 id가 존재하지 않습니다."},"data":null}

kayoung0 committed:
Merge dev into feature/token and resolve conflicts

kayoung0 committed:
fix: 토큰이 존재하지 않을 때 발생하는 예외에 메시지 추가

kayoung0 committed:
fix: application.yml에서 profiles active local 삭제

kayoung0 committed:
chore: application.yml EOF 추가

chanwooDev reviewed (approved):
수고했어요 👍
